File: backtracking/n_queens.go
package main

import "fmt"

func isSafe(board [][]int, row, col int) bool {
	n := len(board)

	// Check column
	for i := 0; i < row; i++ {
		if board[i][col] == 1 {
			return false
		}
	}

	// Check upper left diagonal
	for i, j := row, col; i >= 0 && j >= 0; i, j = i-1, j-1 {
		if board[i][j] == 1 {
			return false
		}
	}

	// Check upper right diagonal
	for i, j := row, col; i >= 0 && j < n; i, j = i-1, j+1 {
		if board[i][j] == 1 {
			return false
		}
	}

	return true
}

func printBoard(board [][]int) {
	for _, row := range board {
		for _, cell := range row {
			if cell == 1 {
				fmt.Print("Q ")
			} else {
				fmt.Print(". ")
			}
		}
		fmt.Println()
	}
	fmt.Println()
}

func solve(board [][]int, row int, solutions *int) {
	if row >= len(board) {
		*solutions++
		printBoard(board)
		return
	}

	for col := 0; col < len(board); col++ {
		if isSafe(board, row, col) {
			board[row][col] = 1
			solve(board, row+1, solutions)
			board[row][col] = 0 // Backtrack
		}
	}
}

func main() {
	n := 4
	board := make([][]int, n)
	for i := range board {
		board[i] = make([]int, n)
	}
	solutions := 0
	solve(board, 0, &solutions)
	fmt.Printf("Total solutions: %d\n", solutions)
}

File: backtracking/n_queens.cpp
#include <iostream>
#include <vector>

bool isSafe(const std::vector<std::vector<int>> &board, int row, int col) {
  int n = board.size();

  // Check this column on upper side
  for (int i = 0; i < row; i++)
    if (board[i][col])
      return false;

  // Check upper diagonal on left side
  for (int i = row, j = col; i >= 0 && j >= 0; i--, j--)
    if (board[i][j])
      return false;

  // Check upper diagonal on right side
  for (int i = row, j = col; i >= 0 && j < n; i--, j++)
    if (board[i][j])
      return false;

  return true;
}

void printBoard(const std::vector<std::vector<int>> &board) {
  for (const auto &row : board) {
    for (int cell : row) {
      std::cout << (cell ? "Q " : ". ");
    }
    std::cout << "\n";
  }
  std::cout << "\n";
}

void solve(std::vector<std::vector<int>> &board, int row, int &solutions) {
  if (row >= board.size()) {
    solutions++;
    printBoard(board);
    return;
  }

  for (int col = 0; col < board.size(); col++) {
    if (isSafe(board, row, col)) {
      board[row][col] = 1;
      solve(board, row + 1, solutions);
      board[row][col] = 0; // Backtrack
    }
  }
}

int main() {
  int n = 4;
  std::vector<std::vector<int>> board(n, std::vector<int>(n, 0));
  int solutions = 0;

  solve(board, 0, solutions);
  std::cout << "Total solutions: " << solutions << std::endl;
  return 0;
}
